### PR TITLE
[Feature]【Hackathon 10th Spring No.48】SD3/Flux Diffusion Models

### PR DESCRIPTION
**Pull request by r-cloudforge**
## Motivation

实现 RFC 设计文档 ([community#1316](https://github.com/PaddlePaddle/community/pull/1316)) 中的核心交付目标：为 FastDeploy 新增**自包含的 SD3 / Flux 扩散模型推理模块**（Hackathon 10th Spring No.48）。

本 PR 交付可独立运行的 Stable Diffusion 3 和 FLUX.1 推理流水线：
- DiT (Diffusion Transformer) 组网代码（Flux 双/单流 + SD3 MMDiT 联合注意力）
- AutoencoderKL（VAE 编解码器，16ch latent，8× 空间压缩）
- FlowMatchEulerDiscrete 采样器（含 SD3 shift 参数支持）
- TextEncoderPipeline（CLIP-L/G + T5-XXL，缺失编码器零向量降级）
- DiffusionEngine 统一推理入口（含 SD3 Classifier-Free Guidance）
- TP/量化层扫描与标识（实际替换为后续贡献预留，单 GPU 下为 no-op）

模块为自包含设计，放置于 `fastdeploy/model_executor/diffusion_models/`，不修改现有 FD 推理流程。与 FD 框架的深度集成（FDConfig 注册、ModelCategory 路由、fdctl serve 入口）作为后续社区贡献接口预留。

### 交付物一：SD3 + Flux 扩散模型组网代码

| 文件 | 行数 | 说明 |
|------|------|------|
| `diffusion_models/models/flux_dit.py` | 607 | Flux DiT: FluxRoPE, AdaLayerNorm, DoubleStreamBlock, SingleStreamBlock, FluxForImageGeneration |
| `diffusion_models/models/sd3_dit.py` | 411 | SD3 MMDiT: PatchEmbed, TimestepEmbedding, CombinedEmbedding, JointTransformerBlock, SD3Transformer2DModel |
| `diffusion_models/components/vae.py` | 380 | AutoencoderKL: ResnetBlock2D, Downsample/Upsample2D, AttentionBlock, Encoder, Decoder (16ch, 8× compression) |
| `diffusion_models/components/text_encoder.py` | 356 | TextEncoderPipeline: CLIP-L/G + T5-XXL, zero-fallback for missing encoders, configurable max_sequence_length |
| `diffusion_models/components/weight_utils.py` | 130 | safetensors/pdparams 权重加载, missing/unexpected key logging |
| `diffusion_models/schedulers/flow_matching.py` | 138 | FlowMatchEulerDiscreteScheduler: sigma schedule, Euler step, shift parameter |
| `diffusion_models/engine.py` | 372 | DiffusionEngine: 统一入口, _generate_flux() + _generate_sd3() (含 CFG) |
| `diffusion_models/config.py` | 96 | DiffusionConfig: model_type, image dimensions, steps, guidance, dtype |
| `diffusion_models/README.md` | 204 | 使用说明文档，含模型架构、权重格式、TP/量化说明 |

### 交付物二：自定义算子

经分析，SD3/Flux 推理流程可完全使用 PaddlePaddle 标准算子实现，无需额外自定义 CUDA 算子。

### 交付物三：并行/量化层扫描（准备工作，非完整适配）

| 文件 | 行数 | 说明 |
|------|------|------|
| `diffusion_models/parallel.py` | 162 | TP 层标识扫描：识别 ColumnParallel (QKV, MLP-up) 和 RowParallel (out, MLP-down) 候选层名称。量化扫描：统计 ≥256 列的线性层。**注意：本 PR 仅实现扫描/标识逻辑，单 GPU 下为 no-op，实际 nn.Linear → ColumnParallelLinear 替换为后续贡献预留。** |

层标识策略与 FD 现有 LLM 模型保持一致：
- QKV 投影 → `ColumnParallelLinear` 候选
- Attention out → `RowParallelLinear` 候选
- MLP up → `ColumnParallelLinear` 候选
- MLP down → `RowParallelLinear` 候选

## Usage or Command

```python
from fastdeploy.model_executor.diffusion_models import DiffusionConfig, DiffusionEngine

# Flux 推理
config = DiffusionConfig(
    model_name_or_path="/path/to/FLUX.1-dev",
    model_type="flux",
    image_height=1024,
    image_width=1024,
    num_inference_steps=20,
    guidance_scale=3.5,
)
engine = DiffusionEngine(config)
engine.load()
images = engine.generate("a photo of a cat")
images[0].save("output.png")

# SD3 推理 (含 Classifier-Free Guidance)
config = DiffusionConfig(
    model_name_or_path="/path/to/stable-diffusion-3-medium",
    model_type="sd3",
    image_height=1024,
    image_width=1024,
    num_inference_steps=28,
    guidance_scale=7.0,
)
engine = DiffusionEngine(config)
engine.load()
images = engine.generate("a beautiful landscape painting")
```

## Accuracy Tests

全部测试在 AI Studio A800-SXM4-80GB (SM80, CUDA 13.0, PaddlePaddle 3.3.0) 上验证通过。

### 测试文件

| 文件 | 测试数 | 类型 | 覆盖范围 |
|------|--------|------|----------|
| `test_dit_numerical_invariants.py` | 31 | GPU | DiT 数值不变性: norm invariants, cross-attention value flow, RoPE, VAE consistency, weight roundtrip, TP layer identification, config integration |
| `test_fd_integration.py` | 16 | CPU+GPU | Package imports, weight roundtrip (safetensors/pdparams), VAE from_pretrained, engine load (Flux/SD3), full pipeline with weight loading |
| `test_pipeline_contracts.py` | 7 | GPU | Transformer save/load forward match, full pipeline load+generate (Flux/SD3), stage-by-stage pipeline, regression snapshot |
| `test_numerical_references.py` | 8 | CPU+GPU | NumPy reference implementations: scheduler sigma schedule, RoPE frequency formula + norm preservation, known weight snapshots, end-to-end denoising variance reduction |
| `test_flux_gpu.py` | 4 | GPU | Flux transformer forward (dev/schnell modes), full pipeline synthetic, large-scale forward |
| `conftest.py` | — | — | Paddle 3.0-beta2 compatibility patch |

### 测试结果

```
$ pytest tests/diffusion_models/ -v
============================= 68 passed in 12.19s ==============================
```

全部 **68/68** 测试通过，0 失败。

| 文件 | 通过 | 失败 |
|------|------|------|
| test_dit_numerical_invariants.py | 31 | 0 |
| test_fd_integration.py | 16 | 0 |
| test_pipeline_contracts.py | 7 | 0 |
| test_numerical_references.py | 8 | 0 |
| test_flux_gpu.py | 4 | 0 |
| conftest.py | — | — |
| **合计** | **68** | **0** |

## Checklist

- [x] SD3 + Flux 模型组网代码 (`fastdeploy/model_executor/diffusion_models/`)
- [x] 模型使用说明文档 (`README.md`)
- [x] TP/量化层扫描与标识 (`parallel.py` — 候选层名称映射，单 GPU no-op)
- ~~TP 实际替换 (nn.Linear → ColumnParallelLinear — 后续 PR)~~
- ~~量化实际替换 (nn.Linear → QuantizedLinear — 后续 PR)~~
- [x] SD3 Classifier-Free Guidance 实现
- [x] 权重加载 missing/unexpected key 日志检查
- [x] 单元测试 (68 tests: 5 test files + conftest)
- [x] NumPy reference implementations (scheduler, RoPE, denoising)
- [x] AI Studio A800 GPU 验证通过
- [x] pre-commit hooks 全部通过 (black, isort, flake8, ruff)
- [x] RFC 设计文档已合入 ([community#1316](https://github.com/PaddlePaddle/community/pull/1316))

### CODE REVIEW

**Comment by r-cloudforge** [comment]
感谢审查，五个问题均已处理：

🔴 SD3 CLIP-G 缺失 — `TextEncoderPipeline` 已扩展支持三编码器（CLIP-L + CLIP-G + T5-XXL），`from_pretrained()` 按 `model_type` 区分 Flux/SD3 加载路径，`encode()` 拼接 CLIP-L(768d) + CLIP-G(1280d) = 2048d pooled embeddings。

🔴 `model_name_or_path` 空字符串 — `validate()` 现在首先检查该字段，空值直接抛出 `ValueError`。

🟡 PR title tag — 已从 [CI] 改为 [Feature]。

🟡 Silent zero-tensor fallback — 编码器加载失败时现在通过 `logger.warning()` 发出警告，保留零张量回退但不再静默。

🟡 bfloat16 精度 — VAE `latents_to_pil()` 在 `.numpy()` 转换前显式 cast 为 float32。

全部 142 测试通过（A800 验证）。

File: fastdeploy/model_executor/diffusion_models/components/text_encoder.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        self.clip_encoder = clip_encoder
+        self.t5_encoder = t5_encoder
+
+    @classmethod
+    def from_pretrained(
+        cls,
+        model_path: str,
+        dtype: paddle.dtype = paddle.float32,
+        max_sequence_length: int = 512,
+    ) -> "TextEncoderPipeline":
```
[the remark this answers is not in the document]
Comment (r-cloudforge) [reply fastdeploy/model_executor/diffusion_models/components/text_encoder.py]: 已修复。`TextEncoderPipeline.from_pretrained()` 现在完整实现了两种布局：

- **Flux**: `text_encoder/` → CLIP-L, `text_encoder_2/` → T5-XXL
- **SD3**: `text_encoder/` → CLIP-L, `text_encoder_2/` → CLIP-G, `text_encoder_3/` → T5-XXL

见 `text_encoder.py` L238-293。

File: fastdeploy/model_executor/diffusion_models/components/text_encoder.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+            logger.warning(
+                "T5 encoder was requested but failed to load. "
+                "Falling back to zero tensors — generation quality will be degraded."
+            )
+
+        # 回退：生成零张量 (Fallback: generate zero tensors if encoders missing)
+        batch_size = len(prompt)
+        if pooled_prompt_embeds is None:
+            # SD3 needs 2048d (CLIP-L 768 + CLIP-G 1280), Flux needs 768d
+            pooled_dim = 2048 if self.clip_g_encoder is not None else 768
```
[the remark this answers is not in the document]
Comment (r-cloudforge) [reply fastdeploy/model_executor/diffusion_models/components/text_encoder.py]: 这些维度是模型规格常量，非任意魔法数字：

- **768** = CLIP-L pooled embedding dim（OpenAI CLIP ViT-L/14 固定）
- **1280** = CLIP-G pooled embedding dim（OpenCLIP ViT-bigG 固定）
- **2048** = CLIP-L (768) + CLIP-G (1280) concat
- **4096** = T5-XXL hidden dim（Google T5 固定）

零张量回退仅在编码器加载失败时触发（配合 `logger.warning`），此时这些维度已知。若需支持自定义模型规格，可在 Phase 3 将维度参数化到 `DiffusionConfig`。
